skimage/segmentation/slic_superpixels.py

```diff
@@ -1,17 +1,119 @@
-from collections.abc import Iterable
+import warnings
+import functools
+import collections as coll
 import numpy as np
 from scipy import ndimage as ndi
+from scipy.spatial.distance import pdist, squareform
 
+from ._slic import (_slic_cython, _enforce_label_connectivity_cython)
 from ..util import img_as_float, regular_grid
-from ..segmentation._slic import (_slic_cython,
-                                  _enforce_label_connectivity_cython)
 from ..color import rgb2lab
 
 
+def _get_mask_centroids(mask, n_centroids, spacing=None):
```

Comment:
Where did you find this technique?  Are you sure it's close to somewhat optimal?  Have you investigated Lloyd's algorithm and others?

Reply:
This is the strategy used in maskSLIC for initializing the centroids. I did not made more investigation but you are right, K-means clustering approaches may be appliyed here also...

Reply:
@stefanv, I made some tests, and `scipy.cluster.vq.kmeans2` is an excellent alternative, thank you again for this suggestion. I will come with some more experimental results later ;-)

Reply:
Nice!

Reply:
My investigation on this topic exposed a bug in the current strategy. I ran the following lines:
```python

In [1]: from skimage.segmentation.slic_superpixels import _get_mask_centroids                                          

In [2]: import numpy as np                                                                                             

In [3]: mask = np.zeros((1, 100, 100), dtype=bool)                                                                     

In [4]: mask[0, 20:60, 30:90] = True                                                                                   

In [5]: c, _ = _get_mask_centroids(mask, 100)

In [6]: P.imshow(mask[0])                                                                                              
Out[6]: <matplotlib.image.AxesImage at 0x7f14e803b150>

In [7]: P.plot(c[:, 2], c[:, 1], 'rx')                                                                                 
Out[7]: [<matplotlib.lines.Line2D at 0x7f14e8e81390>]
```

With the current implementation I obtained
![bug](https://user-images.githubusercontent.com/3438227/79075637-65ccb380-7cf4-11ea-99d8-804aa1d180d4.png)
I fixed the bug, here is what I obtained:
![fix](https://user-images.githubusercontent.com/3438227/79075745-f4413500-7cf4-11ea-940a-94b8f3b26b0a.png)
Not really good :slightly_frowning_face:!
I implemented the suggestion of @stefanv, and here is the result:
![kmeans](https://user-images.githubusercontent.com/3438227/79075788-2a7eb480-7cf5-11ea-89e2-35e7eb2b59ba.png)
Much better :slightly_smiling_face: !
Moreover, the timing performances are improved:
- old:
```python
In [5]: %timeit _get_mask_centroids(mask, 100)                                                                         
17.9 ms ± 192 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
- new:
```python
In [8]: %timeit _get_mask_centroids(mask, 100)                                                                         
14.2 ms ± 107 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```